File: ruby/run-length-encoding/run_length_encoding.rb
module BookKeeping
  VERSION = 2
end

class RunLengthEncoding

  def self.encode(input)
    input.chars.chunk { |chars| chars }.inject("") do |result, (name, chars)|
      value = chars.size == 1 ? chars[0] : chars.size.to_s + name
      result << value
    end
  end

  def self.decode(input)
    input.scan(/(\d+)?(.)/).inject("") do |result, (count, char)|
      value = count ? char * count.to_i : char
      result << value
    end
  end

end

File: python/run-length-encoding/run_length_encoding.py
from itertools import groupby

def decode(s):
    result = ''
    c_buffer = ''
    for c in s:
        if c.isdigit():
            c_buffer += c
        else:
            result += c * int(c_buffer or 1)
            c_buffer = ''
    return result


def encode(s):
    result = ''
    for v, grp in groupby(s):
        c = len(''.join(grp))
        if c == 1:
            result += v
        else:
            result += str(c) + v
    return result
